File: RxTools-library/src/main/java/com/vondear/rxtools/view/colorpicker/renderer/SimpleColorWheelRenderer.java
package com.vondear.rxtools.view.colorpicker.renderer;

import android.graphics.Color;
import android.graphics.Paint;

import com.vondear.rxtools.view.colorpicker.ColorCircle;
import com.vondear.rxtools.view.colorpicker.builder.PaintBuilder;

public class SimpleColorWheelRenderer extends AbsColorWheelRenderer {
    private Paint selectorFill = PaintBuilder.newPaint().build();
    private float[] hsv = new float[3];

    @Override
    public void draw() {
        final int setSize = colorCircleList.size();
        int currentCount = 0;
        float half = colorWheelRenderOption.targetCanvas.getWidth() / 2f;
        int density = colorWheelRenderOption.density;
        float maxRadius = colorWheelRenderOption.maxRadius;

        for (int i = 0; i < density; i++) {
            float p = (float) i / (density - 1); // 0~1
            float radius = maxRadius * p;
            float size = colorWheelRenderOption.cSize;
            int total = calcTotalCount(radius, size);

            for (int j = 0; j < total; j++) {
                double angle = Math.PI * 2 * j / total + (Math.PI / total) * ((i + 1) % 2);
                float x = half + (float) (radius * Math.cos(angle));
                float y = half + (float) (radius * Math.sin(angle));
                hsv[0] = (float) (angle * 180 / Math.PI);
                hsv[1] = radius / maxRadius;
                hsv[2] = colorWheelRenderOption.lightness;
                selectorFill.setColor(Color.HSVToColor(hsv));
                selectorFill.setAlpha(getAlphaValueAsInt());

                colorWheelRenderOption.targetCanvas.drawCircle(x, y, size - colorWheelRenderOption.strokeWidth, selectorFill);

                if (currentCount >= setSize)
                    colorCircleList.add(new ColorCircle(x, y, hsv));
                else colorCircleList.get(currentCount).set(x, y, hsv);
                currentCount++;
            }
        }
    }
}

File: RxUI/src/main/java/com/tamsiree/rxui/view/colorpicker/renderer/SimpleColorWheelRenderer.kt
package com.tamsiree.rxui.view.colorpicker.renderer

import android.graphics.Color
import com.tamsiree.rxui.view.colorpicker.ColorCircle
import com.tamsiree.rxui.view.colorpicker.builder.PaintBuilder

/**
 * @author tamsiree
 * @date 2018/6/11 11:36:40 整合修改
 */
class SimpleColorWheelRenderer : AbsColorWheelRenderer() {
    private val selectorFill = PaintBuilder.newPaint().build()
    private val hsv = FloatArray(3)
    override fun draw() {
        val setSize = colorCircleList.size
        var currentCount = 0
        val half = colorWheelRenderOption?.targetCanvas?.width!! / 2f
        val density = colorWheelRenderOption?.density
        val maxRadius = colorWheelRenderOption?.maxRadius
        for (i in 0 until density!!) {
            // 0~1
            val p = i.toFloat() / (density - 1)
            val radius = maxRadius!! * p
            val size = colorWheelRenderOption?.cSize
            val total = calcTotalCount(radius, size!!)
            for (j in 0 until total) {
                val angle = Math.PI * 2 * j / total + Math.PI / total * ((i + 1) % 2)
                val x = half + (radius * Math.cos(angle)).toFloat()
                val y = half + (radius * Math.sin(angle)).toFloat()
                hsv[0] = (angle * 180 / Math.PI).toFloat()
                hsv[1] = radius / maxRadius
                hsv[2] = colorWheelRenderOption?.lightness!!
                selectorFill.color = Color.HSVToColor(hsv)
                selectorFill.alpha = alphaValueAsInt
                colorWheelRenderOption?.targetCanvas?.drawCircle(x, y, size - colorWheelRenderOption?.strokeWidth!!, selectorFill)
                if (currentCount >= setSize) {
                    colorCircleList.add(ColorCircle(x, y, hsv))
                } else {
                    colorCircleList[currentCount][x, y] = hsv
                }
                currentCount++
            }
        }
    }
}